packages/input-manager/src/integrations/midi/index.ts

```diff
@@ -0,0 +1,125 @@
+import { Input, Output } from 'easymidi'
+import { Logger } from '../../logger'
+import { Device } from '../../devices/device'
+import { Symbols } from '../../lib'
+import { SomeFeedback } from '../../feedback/feedback'
+
+export interface MIDIDeviceConfig {
+	inputName: string
+	outputName?: string
+}
+
+export class MIDIDevice extends Device {
+	#input: Input | undefined
+	#output: Output | undefined
+	#config: MIDIDeviceConfig
+	#feedbacks: Record<number, SomeFeedback> = {}
+
+	constructor(config: MIDIDeviceConfig, logger: Logger) {
+		super(logger)
+		this.#config = config
+	}
+
+	async init(): Promise<void> {
+		this.#input = new Input(this.#config.inputName)
+		if (this.#config.outputName) this.#output = new Output(this.#config.outputName)
+		this.#input.on('noteon', (msg) => {
+			const triggerId = `${msg.channel}_${msg.note} ${Symbols.DOWN}`
+			this.emit('trigger', {
+				triggerId,
+				arguments: {
+					velocity: msg.velocity,
+				},
+			})
+			this.updateFeedback(msg.note).catch(console.error)
+		})
+		this.#input.on('noteoff', (msg) => {
+			const triggerId = `${msg.channel}_${msg.note} ${Symbols.UP}`
+			this.emit('trigger', {
+				triggerId,
+				arguments: {
+					velocity: msg.velocity,
+				},
+			})
+			this.updateFeedback(msg.note).catch(console.error)
+		})
+		this.#input.on('cc', (msg) => {
+			const triggerId = `${msg.channel}_${msg.controller} cc`
+			this.emit('trigger', {
+				triggerId,
+				arguments: {
+					value: msg.value,
+				},
+				replacesPrevious: true,
+			})
+		})
+	}
+
+	async destroy(): Promise<void> {
+		await super.destroy()
+		if (this.#input) this.#input.close()
+		if (this.#output) this.#output.close()
+	}
+
+	private static parseTriggerId(triggerId: string): { note: number; isNote: boolean; isCC: boolean; isUp: boolean } {
+		const triggerElements = triggerId.match(/(\d+)_(\d+)\s+(\S+)/)
+		if (!triggerElements) return { note: 0, isNote: false, isCC: false, isUp: false }
+		const note = Number.parseInt(triggerElements[2] ?? '0')
+		const isNote = triggerElements[3] === Symbols.UP || triggerElements[3] === Symbols.DOWN
+		const isCC = triggerElements[3] === 'cc'
+		const isUp = triggerElements[3] === Symbols.UP
+		return { note, isNote, isCC, isUp }
+	}
+
+	private selectOutputNote(note: number): number {
+		if (note >= 8 && note <= 23) {
```

Comment:
I wonder how portable this is, or whether this is specific to your midi device. 
Ill test with something else at some point

Reply:
No, this is completely not portable. Unfortunately, the MIDI device feedback world seems to be completely non-standard, so this will need to be provided through configuration, but with the configuration system missing, this is simply a quick placeholder.